packages/pigeon/example/app/linux/messages.g.cc

```diff
@@ -0,0 +1,548 @@
+    return;
+  }
+
+  if (fl_value_get_type(message) != FL_VALUE_TYPE_LIST ||
+      fl_value_get_length(message) != 1 ||
+      fl_value_get_type(fl_value_get_list_value(message, 0)) !=
+          FL_VALUE_TYPE_CUSTOM) {
+    return;
+  }
+
+  self->vtable->send_message(self,
+                             MY_MESSAGE_DATA(fl_value_get_custom_value_object(
+                                 fl_value_get_list_value(message, 0))),
+                             response_handle, self->user_data);
+}
+
+static void my_example_host_api_dispose(GObject* object) {
+  MyExampleHostApi* self = MY_EXAMPLE_HOST_API(object);
+  g_clear_object(&self->messenger);
+  if (self->user_data != nullptr) {
+    self->user_data_free_func(self->user_data);
+  }
+  self->user_data = nullptr;
+
+  g_clear_object(&self->get_host_language_channel);
+  g_clear_object(&self->add_channel);
+  g_clear_object(&self->send_message_channel);
+  G_OBJECT_CLASS(my_example_host_api_parent_class)->dispose(object);
+}
+
+static void my_example_host_api_init(MyExampleHostApi* self) {}
+
+static void my_example_host_api_class_init(MyExampleHostApiClass* klass) {
+  G_OBJECT_CLASS(klass)->dispose = my_example_host_api_dispose;
+}
+
+MyExampleHostApi* my_example_host_api_new(FlBinaryMessenger* messenger,
+                                          const MyExampleHostApiVTable* vtable,
+                                          gpointer user_data,
+                                          GDestroyNotify user_data_free_func) {
+  MyExampleHostApi* self = MY_EXAMPLE_HOST_API(
+      g_object_new(my_example_host_api_get_type(), nullptr));
+  self->messenger = g_object_ref(messenger);
+  self->vtable = vtable;
+  self->user_data = user_data;
+  self->user_data_free_func = user_data_free_func;
+
+  g_autoptr(MyExampleHostApiCodec) codec = my_example_host_api_codec_new();
+  self->get_host_language_channel =
+      fl_basic_message_channel_new(messenger,
+                                   "dev.flutter.pigeon.pigeon_example_package."
+                                   "ExampleHostApi.getHostLanguage",
+                                   FL_MESSAGE_CODEC(codec));
+  fl_basic_message_channel_set_message_handler(
+      self->get_host_language_channel, get_host_language_cb, self, nullptr);
+  self->add_channel = fl_basic_message_channel_new(
+      messenger, "dev.flutter.pigeon.pigeon_example_package.ExampleHostApi.add",
+      FL_MESSAGE_CODEC(codec));
+  fl_basic_message_channel_set_message_handler(self->add_channel, add_cb, self,
+                                               nullptr);
+  self->send_message_channel = fl_basic_message_channel_new(
+      messenger,
+      "dev.flutter.pigeon.pigeon_example_package.ExampleHostApi.sendMessage",
+      FL_MESSAGE_CODEC(codec));
+  fl_basic_message_channel_set_message_handler(self->send_message_channel,
+                                               send_message_cb, self, nullptr);
```

Comment:
Nothing is leaked, because the channels are owned by the api object. If that object is unreferenced then all the channels are closed. If the engine is unreferenced before, the channels will stop working but the objects will remain until the api object is unreferenced.

In this case it looks something like this:
```c
static PigeonExamplePackageExampleHostApiGetHostLanguageResponse* get_host_language(PigeonExamplePackageExampleHostApi* api, gpointer user_data) {
   // ...
}

static PigeonExamplePackageExampleHostApiVTable vtable = {
   .get_host_language = get_host_language
};

int main(int argc, char** argv) {
  g_autoptr(PigeonExamplePackageExampleHostApi) api = pigeon_example_package_example_host_api_new(messenger, suffix, &vtable, nullptr, nullptr);
}
```

Is there a case where it's difficult to keep the api object alive for the lifetime of the app (or at least the lifetime that you want to use that API)?

Reply:
> Nothing is leaked, because the channels are owned by the api object.

Sorry, I confused myself between what the example is doing and what I wanted to do in terms of which references are strong. The example app wasn't the right reference because the app doesn't have the same lifetime structure plugins do.

> Is there a case where it's difficult to keep the api object alive for the lifetime of the app (or at least the lifetime that you want to use that API)?

Normally the thing that keeps a plugin instance alive is the handler ([example](https://github.com/flutter/packages/blob/0e3460cba24784314d1374603856b9629e75aec5/packages/url_launcher/url_launcher_linux/linux/url_launcher_plugin.cc#L129-L130)). Because the handler is freed when the engine is destroyed, that means that the plugin lifetime is managed by the engine lifetime, which is what we want.

But I don't see how to replicate that with this Pigeon-generated code. It looks like all of the message handler references it creates are weak, which means there's no mechanism to have the engine own the plugin any more (other than creating a bogus channel manually in the plugin code for the sole purpose of giving the engine an owning ref, but that would clearly be a hack).

The cycle I was thinking of in my previous comment was this one from my initial comment:
>> But if I naively port that pattern over to my_example_host_api_new, there would be a cycle, because the plugin instance would own the MyExampleHostApi instance, which would in turn own the plugin instance, and because this is passing nullptr nothing ever gets cleaned up.

I.e., using only Pigeon (without the hack I mentioned just above), it seems like either I'll have the plugin immediately destroyed (if I give the API object a weak ref to the plugin), or I'll have a cycle (if I give the API object a strong reference to the plugin).

Reply:
I can change it to make the API object behave like a plugin, i.e. there's no need to keep a reference to it, and it is destroyed when the channels are closed. You lose the means to explicitly close the channels, but this is not required?

Reply:
Some plugins do need the ability to close the channels, usually plugins that have instance objects that are created and destroyed (e.g., video player or webveiw, where each player or view has a corresponding object with its own channel).

The way this works on the other platforms is that the channel setup (which is just a function, not an object) can be called with a `null` handler, which in turn sets all the channel registrations to null handlers.

Reply:
I've removed the API object and made a set_method_handlers/clear_method_handlers functions as a replacement. I think this should do meet the requirements.